gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-sync/src/main/java/io/gravitee/gateway/services/sync/process/kubernetes/fetcher/ConfigMapEventFetcher.java

```diff
@@ -42,25 +46,52 @@
 @RequiredArgsConstructor
 public class ConfigMapEventFetcher {
 
-    private static final String LABEL_MANAGED_BY = "managed-by";
-    private static final String LABEL_GIO_TYPE = "gio-type";
-    protected static final String GRAVITEE_IO = "gravitee.io";
-    protected static final String API_DEFINITION_V1_ALPHA1 = "v1alpha1";
-    protected static final String APIDEFINITIONS_TYPE = "apidefinitions.gravitee.io";
+    public static final String API_DEFINITIONS_KIND = "apidefinitions.gravitee.io";
+
+    protected static final String RESOURCE_GROUP = "gravitee.io";
+    protected static final String RESOURCE_VERSION = "v1alpha1";
+
     protected static final String DATA_ENVIRONMENT_ID = "environmentId";
     protected static final String DATA_API_DEFINITION_VERSION = "apiDefinitionVersion";
     protected static final String DATA_DEFINITION = "definition";
+
+    private static final String LABEL_MANAGED_BY = "managed-by";
+    private static final String LABEL_GIO_TYPE = "gio-type";
+
     private static final int RETRY_DELAY_MILLIS = 10000;
+
+    private static final String EVENT_ADDED = "ADDED";
+    private static final String EVENT_MODIFIED = "MODIFIED";
+    private static final String EVENT_DELETED = "DELETED";
+
     private final KubernetesClient client;
     private final String[] namespaces;
     private final ObjectMapper objectMapper;
 
+    private final Map<String, String> resourceVersions = new ConcurrentHashMap<>();
+
     public int bulkEvents() {
         return 1;
     }
 
-    public Flowable<List<io.gravitee.repository.management.model.Event>> fetchLatest() {
-        return watchConfigMaps()
+    public Flowable<List<io.gravitee.repository.management.model.Event>> fetchAll(String kind) {
+        return listConfigMaps(kind)
+            .doOnError(err -> {
+                if (err instanceof ResourceVersionNotFoundException) {
+                    resourceVersions.remove(kind);
+                }
```

Comment:
I probably missed something but I don't understand why it is not handled directly in the listConfigMaps method considering the put is? Another point is that I'm not sure why a removal is needed in case of an exception thrown as I would expect the put isn't made?

Reply:
@jhaeyaert 

This exception is raised only for [410 GONE responses](https://github.com/gravitee-io/gravitee-kubernetes/pull/76/files#diff-cfaf8e67596c47ac2303792be45f147c2be301fe826536604636f5dd59e93bd9R146) from the kubernetes API server.

When we retry, the cluster might not be aware anymore of the resource version we stored and return this status code. The idea is to not retry for ever because we address a wrong request to the API server.

This is the extract from the kubernetes documentation

> A given Kubernetes server will only preserve a historical record of changes for a limited time. Clusters using etcd 3 preserve changes in the last 5 minutes by default. When the requested watch operations fail because the historical version of that resource is not available, clients must handle the case by recognizing the status code 410 Gone, clearing their local cache, performing a new get or list operation, and starting the watch from the resourceVersion that was returned.